### ISSUE TITLE
Docker Agenda/Task list

### ISSUE THREAD

**Issue by gundalow** [ansible/community#408]
## How to help

* Firstly, click the `Subscribe` button on the right hand side so you will see update.
* Ansible used GitHub labels to categorize Issues and PRs, therefore everything Docker related has [/label/docker](https://github.com/ansible/ansible/labels/docker)
* Reviewing [Docker PRs](https://github.com/ansible/ansible/pulls?q=is%3Aopen+label%3Adocker+is%3Apr) is very useful

  * Actually using and testing the PRs is one of the best types of feedback we can get. Add a comment saying what you tested and how it worked: [example of good testing feedback](https://github.com/ansible/ansible/pull/46151#issuecomment-435026764)
* Adding and extending the integration tests
* What else?

## How to speak to us

* Feel free to add comments on this issue
* Join us in `#ansible-devel` on Libera - [Ansible IRC channels](https://docs.ansible.com/ansible/latest/community/communication.html)

**Comment by felixfontein** [comment]
I've compiled a list of PRs which aren't from team docker contributors (i.e. all except ansible/ansible#49235 and ansible/ansible#48574). Most of them look/are dead, except two docker inventory scripts.

### docker_container
Both of them look dead. The first one isn't urgend anymore (since docker has been deprecated for a while and docker_container around for a while), but the PR seems to have problems; there are review comments from @chouseknecht which have never been incorporated. I guess closing it is fine, as ansible/ansible#19103 still reminds of it. The second is about adding `--cpu` support. The main problem with it is that the interface isn't so great. It would be nice to have an easy way to specify CPU usage which works with different docker-py / docker API versions. This is tracked in ansible/ansible#34320, and there's a related issue ansible/ansible#49008 for `--cpu-rt-runtime`.
- https://github.com/ansible/ansible/pull/19105 [dead]
- https://github.com/ansible/ansible/pull/38063 [essentially dead, redesign required]

### docker_image
There is one PR about adding a `network` parameter, which looks good; it is missing some backwards compatibility stuff, and a changelog entry. If the author continues to not reply, I suggest we close this and recreate it. Anyone interested in handling this one?
- https://github.com/ansible/ansible/pull/45127 [probably dead]

### docker_network
There are two PRs for adding two features (`attachable` and `scope`), both have conflicts and seem dead, but it should be simple to create a new PR adding both features. @DBendit are you interested in working on this?
- https://github.com/ansible/ansible/pull/35288 [probably dead]
- https://github.com/ansible/ansible/pull/45552 [probably dead]

### docker_service
The docker module with the most amount of issues... There's one PR fixing something, but it has conflicts and might be dead. Is anyone interested working on this PR and/or the docker_service module in general? There's a *lot* to do here.
- https://github.com/ansible/ansible/pull/31305 [maybe dead]

### docker connection plugin
There are two PRs for the docker connection plugin which seem to be dead. The first one is very old, and the author stated that he won't work on it anymore (i.e. definitely dead). It will require some work to rebase this, maybe it's easier to rewrite it. The second one is rather new, but the author never responded after the initial two commits. (It was created during a hackathon, @pilou- knows more.) From the looks of it, it's probably also dead. Is anyone interested in working on the connection plugin?
- https://github.com/ansible/ansible/pull/16198 [dead]
- https://github.com/ansible/ansible/pull/46547 [probably dead]

### inventory scripts
There are three inventory script related PRs. Two add a new inventory script, the third improves an existing one. AFAIK inventory scripts are dead (there is a [new plugin category](https://docs.ansible.com/ansible/latest/plugins/inventory.html) for this), but since I've never used one I don't know much about this. What shall we do about them?
- docker-machine: https://github.com/ansible/ansible/pull/23149 [dead]
- swarm: https://github.com/ansible/ansible/pull/42667 [alive]
- service/stack for swarm: https://github.com/ansible/ansible/pull/48021 [alive]

@gundalow feel free to close the dead PRs :) For the ones where I only asked a few days ago, maybe we should wait a few more days before closing them?

**Comment by DBendit** [comment]
@felixfontein Yeah, I can take those two issues. I've also got personal to-dos to consolidate that Python boolean handling from https://github.com/ansible/ansible/pull/48105 into `docker_common.py` and to generally expand the `docker_network` integ tests.

**Comment by gundalow** [comment]
@DBendit Thanks for offering to take a look at some of this. Which two issues are you referring to?

**Comment by DBendit** [comment]
@gundalow The two that @felixfontein specifically asked me about, https://github.com/ansible/ansible/pull/35288 and https://github.com/ansible/ansible/pull/45552

**Comment by felixfontein** [comment]
@akshay196 @danihodovic @dariko @DBendit @jwitko @kassiansun @tbouvet

As mentioned above, we currently have a "problem" with `docker_service`. Nobody from us seems to be interested in it, and there seem to be several active users which have problems ([19081](https://github.com/ansible/ansible/issues/19081), [19371](https://github.com/ansible/ansible/issues/19371), [20458](https://github.com/ansible/ansible/issues/20458), [21504](https://github.com/ansible/ansible/issues/21504), [22358](https://github.com/ansible/ansible/issues/22358), [23607](https://github.com/ansible/ansible/issues/23607), [25943](https://github.com/ansible/ansible/issues/25943), [26937](https://github.com/ansible/ansible/issues/26937), [28768](https://github.com/ansible/ansible/issues/28768), [29498](https://github.com/ansible/ansible/issues/29498), [29513](https://github.com/ansible/ansible/issues/29513), [29527](https://github.com/ansible/ansible/issues/29527), [29532](https://github.com/ansible/ansible/issues/29532), [30548](https://github.com/ansible/ansible/issues/30548), [31924](https://github.com/ansible/ansible/issues/31924), [33731](https://github.com/ansible/ansible/issues/33731), [33960](https://github.com/ansible/ansible/issues/33960), [37059](https://github.com/ansible/ansible/issues/37059), [37471](https://github.com/ansible/ansible/issues/37471), [42506](https://github.com/ansible/ansible/issues/42506), [47978](https://github.com/ansible/ansible/issues/47978)) which nobody tried to fix for quite some time (except ansible/ansible#31305, which looks inactive). Is anyone of you interested on working on this?

If not, we should try to look for people interested in the module. I guess the easiest way is to post an "advertisement" in each of these issues, and hope that someone there has interest in improving the situation (or at least working on a specific fix). What do you think about the following text?

> Thank you for your interest in the `docker_service` module. Unfortunately, it is currently essentially unmaintained. If you are interested in maintaining it, or working on this (or other) bugs, please feel free to write that here and start working on it!

**Comment by DBendit** [comment]
@felixfontein Tempted as I am to jump in, I don't know that I can commit to taking over another module until I get a better idea of how it works. I *do* like the idea of soliciting folks reporting these issues to submit PRs at the very least. I think that trying to get issue submitters to commit to ownership of the service is a bit too aggressive, but letting them know that their issue isn't likely to get fixed if they don't fix it themselves might prompt one or two of them to action, and then we can ask if they'd like to keep contributing.

**Comment by felixfontein** [comment]
@WojciechowskiPiotr suggested [here](https://github.com/ansible/ansible/pull/50428#issuecomment-451687999) to rename `docker_service` to `docker_compose` (since that's what it actually is). I like the idea. Are there any other opinions on this?

**Comment by danihodovic** [comment]
I think that's a better name since services are associated with docker swarm, e.g `docker service createhttps://docs.docker.com/engine/reference/commandline/service_create/.

However it's a breaking change and I don't know how Ansible deals with those.

**Comment by felixfontein** [comment]
The standard way to deal with this (AFAIK) is to rename the module, and create a symlink (old module name prepended with `_`) to the new module, and add a warning to the module if it is used with the old module name. (See [here](https://github.com/ansible/ansible/pull/39816/files) for an example.) The symlink will be kept for four versions (regular Ansible deprecation cycle), so there's enough time for people to adjust.

**Comment by felixfontein** [comment]
One disadvantage is that after the rename, ansibot will mark **all** issues and PRs for docker_service as `deprecated`. It might be possible to avoid this by writing `!component =lib/ansible/modules/cloud/docker/docker_compose.py` quickly enough (if one does it too late, I think the `deprecated` won't be removed automatically).

**Comment by felixfontein** [comment]
Here's a list of current PRs which are pretty mature and won't mind some additional eyes/reviews/opinions :)

#50584: big refactoring of docker_swarm by @WojciechowskiPiotr. This adds a new module_utils module `docker_swarm` and a new Ansible module `docker_node`, which contains some of the functionality of `docker_swarm`. Besides this, @WojciechowskiPiotr would like to join $team_docker! :) (See also #50622 for `docker_swarm_facts` which will use this module_utils module, and there will be another PR for `docker_node_facts` once #50584 is merged.)

#51373: adds a new `docker_host_facts` module which allows to list various kind of objects (containers, images, networks, volumes) and query facts about the docker daemon.

#44692: improves the image finding methods; docker has problems finding images when the user prepends `docker.io/` or `docker.io/library/` to an image name (coming from docker hub), which I think is something we should fix. More opinions are needed.

#51145: improves the `force` option of `docker_volume` (split up into two different options to prevent future confusion).

#51192: trying to get rid of regular modules returning facts. We can't simply remove them, since there's no good way to deprecate them (and warn the user if she uses them), but at least we can add an alternative return value.

#51216: this is about changing the default values for `docker_swarm_service`. Some more eyes would be appreciated so we don't accidentally break something.

#51271: looks like the documentation for `DOCKER_TLS` and `DOCKER_TLS_VERIFY` are not correct, and neither is the fact that they are treated as mutually exclusive.

**Comment by jwitko** [comment]
> @WojciechowskiPiotr suggested [here](https://github.com/ansible/ansible/pull/50428#issuecomment-451687999) to rename `docker_service` to `docker_compose` (since that's what it actually is). I like the idea. Are there any other opinions on this?

I am very much in favor of this.  I think renaming it may also help in finding an official maintainer.

**Comment by felixfontein** [comment]
@jwitko @danihodovic it's good that you agree, because we already renamed the module ;-) I guess I should have mentioned that here. The PR was #51035.

**Comment by danihodovic** [comment]
Haha. Sounds good :+1:

**Comment by felixfontein** [comment]
@danihodovic sorry, completely forgot your question :) Yes, it will land in 2.8. The old module name will still work, though you'll get a warning telling you that the module was renamed.

**Comment by WojciechowskiPiotr** [comment]
Useful feature idea in #38859 - copy files(s)/dir(s) to and from a container, so `docker cp` command implementation. Might be a feature in `docker_container` or maybe worth considering it as a new module `docker_container_cp` due to requirements of identity (no compression), gzip, bzip2, xz as the content must be provided as an archive.

**Comment by felixfontein** [comment]
General information: feature freeze for Ansible 2.8 is on March 21st. After that, no more new features are accepted for Ansible 2.8, only bugfixes. (By that date, `stable-2.8` will be branched from `devel` IIRC, and the procedure for updating that one will be similar than for updating `stable-2.7` and before - i.e. no new features.)

So if you plan to work on a new feature, either have it merged until then, or you'll have to wait another 6 months or so :)

(Source: [tentative 2.8 roadmap](https://docs.ansible.com/ansible/devel/roadmap/ROADMAP_2_8.html#expected).)

**Comment by felixfontein** [comment]
About ansible/ansible#38859: that's already possible with the docker connection plugin. With that one, you can connect to an Ansible container and execute modules in it like on any other host. (The issue is about that it's hard to find out how to actually set that up.) In case we want to have explicit support, I'm against adding it to `docker_container`, as that module is already complex enough and this operation is not part of the regular create/start/stop/remove container cycle.

**Comment by WojciechowskiPiotr** [comment]
> About [ansible/ansible#38859](https://github.com/ansible/ansible/issues/38859): that's already possible with the docker connection plugin. With that one, you can connect to an Ansible container and execute modules in it like on any other host. (The issue is about that it's hard to find out how to actually set that up.) In case we want to have explicit support, I'm against adding it to `docker_container`, as that module is already complex enough and this operation is not part of the regular create/start/stop/remove container cycle.

Ok, I tried the plugin (I was not aware of it) and I still think we may consider module for this operation because:

- Plugin requires python installed inside of the destination container (!!!!)
- It expect the container name as an inventory entry provided to playbook, not the hosts. This is quite unusable comparing to how the docker modules for ansible 2.8 are working, and in such approach facts gathered by `docker_*_facts` are useless.
- I was able to copy file if container is run on same host as host executing the playbook.
- I was able to execute it on remote host so far but only when providing the TLS parameters from command line, not from playbook itself.
- Might be that many exceptions are not handled in the plugin code as I experienced traceback quite many times

I agree that huge part of documentation for the plugin is missing but so far (i am not saying this is 100% accurate) it looks for me that there is no interoperability with currently developed docker plugins. 

The Docker API provides `/containers/{id}/archive` GET and POST methods, it requires that content is put into an archive but this is dependecy on host running the module, there is no dependency on the docker host or container itself

**Comment by WojciechowskiPiotr** [comment]
Any opinions on my comment above? Anyone ever used this plugin and can share more details? 
Is there any rule that if we have a connection plugin we cannot have a module doing a similar operation?

**Comment by felixfontein** [comment]
I wanted to collect a list of bigger things we should get done for Ansible 2.8. Here it is:

 - The docker_image module has quite some quirks and is hard to use (#29514, #49909, #44077). Maybe we should deprecate the module and replace it by a set of more specialized modules (docker_image_pull, docker_image_push, docker_image_tag, docker_image_build, docker_image_import; what about deleting images, then?)? Any ideas/opinions/...?

 - Find out what to do with docker_swarm (see #49547; this can be done with docker_node in Ansible 2.8, so what do we want to do with this functionality in docker_swarm?).

 - `docker_node` and `docker_swarm` should be adjusted to work (better, or at all) with older docker-py / docker API versions.

**Comment by gundalow** [comment]
We only have a month till [Core Freeze](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_8.rst) for Ansible 2.8 so we need to be mindful what we want to commit to.

**Comment by WojciechowskiPiotr** [comment]
> * The docker_image module has quite some quirks and is hard to use (#29514, #49909, #44077). Maybe we should deprecate the module and replace it by a set of more specialized modules (docker_image_pull, docker_image_push, docker_image_tag, docker_image_build, docker_image_import; what about deleting images, then?)? Any ideas/opinions/...?

I assume that adding the `action` as suggested in #44077 is not preferred way reflecting the way Ansible should work? In such a case, it might be a good idea to split it into multiple modules but that will be a bigger change as we cannot retire the `docker_image` just like that.
On the other hand currently, there are three states - `absent`, `present` and `build`. The `present` should be used to pull node or add/remove tags in a similar way it is in `docker_swarm_node` now. The `build` is clear, prunning unwanted images feature is covered in `docker_prune` module. So we may just focus on creating the `docker_registry` (or extending the `docker_login`) for pushing image and other registry operations and support more than hub.docker.com in long term

> * Find out what to do with docker_swarm (see #49547; this can be done with docker_node in Ansible 2.8, so what do we want to do with this functionality in docker_swarm?).

If @tbouvet is busy I may find some time to adjust the module to `AnsibleDockerSwarmClient` class first then we may get rid of features handled in `docker_warm_facts` and `docker_node_facts` but I understand that will take 3 Ansible releases to accomplis

> * `docker_node` and `docker_swarm` should be adjusted to work (better, or at all) with older docker-py / docker API versions.

Can you share some more highlight on the problems? Or link the issues numbers.

**Comment by felixfontein** [comment]
> > * The docker_image module has quite some quirks and is hard to use (#29514, #49909, #44077). Maybe we should deprecate the module and replace it by a set of more specialized modules (docker_image_pull, docker_image_push, docker_image_tag, docker_image_build, docker_image_import; what about deleting images, then?)? Any ideas/opinions/...?
> 
> I assume that adding the `action` as suggested in #44077 is not preferred way reflecting the way Ansible should work?

I think you mean ansible/ansible#29514. But yeah, I don't think this is how it should work. I mainly quoted the issue to point out that the module has been confusing for already quite some time ;)

>  In such a case, it might be a good idea to split it into multiple modules but that will be a bigger change as we cannot retire the `docker_image` just like that.
> On the other hand currently, there are three states - `absent`, `present` and `build`. The `present` should be used to pull node or add/remove tags in a similar way it is in `docker_swarm_node` now. The `build` is clear, prunning unwanted images feature is covered in `docker_prune` module.

Ironically, the documentation of `state` says "NOTE: build is DEPRECATED and will be removed in release 2.3. Specifying build will behave the same as present." (2.3 was released in April 2017).

> So we may just focus on creating the `docker_registry` (or extending the `docker_login`) for pushing image and other registry operations and support more than hub.docker.com in long term

Supporting more than hub.docker.com should already work fine. (See the integration tests; they set up a local instance of docker registry and push to it.)

I'll spent some more time thinking about this. Also, if anyone else has some ideas, let's take a look at this again the at the beginning of next week or so.

> > * Find out what to do with docker_swarm (see #49547; this can be done with docker_node in Ansible 2.8, so what do we want to do with this functionality in docker_swarm?).
> 
> If @tbouvet is busy I may find some time to adjust the module to `AnsibleDockerSwarmClient` class first then we may get rid of features handled in `docker_warm_facts` and `docker_node_facts` but I understand that will take 3 Ansible releases to accomplis

Actually, the deprecation cycle is four releases. But having explicit deprecation warnings which inform the user what to use instead is always a good idea :)

@tbouvet what is your schedule? Are you interested on working on this, or would you like @WojciechowskiPiotr or someone else to do this?

> > * `docker_node` and `docker_swarm` should be adjusted to work (better, or at all) with older docker-py / docker API versions.
> 
> Can you share some more highlight on the problems? Or link the issues numbers.

For docker_node there are no issues (yet). For docker_swarm, there's ansible/ansible#51175 (with preliminary PR ansible/ansible#51264). There might be some more problems which nobody noticed (or reported) so far.

What is relatively easy is testing with old versions of docker/docker-py. I've done that extensively for some other modules, like docker_container: first install the earliest docker-py version supported, then run the integration tests. If everything's fine, upgrade to the next minor docker-py release, and repeat. Eventually, continue with docker releases.

That approach needs good enough integration tests, though, and the tests should be prepared to degrade gracefully for older docker-py versions (see https://github.com/ansible/ansible/blob/devel/test/integration/targets/docker_container/tasks/tests/options.yml#L51-L60 for an example).

Testing with older docker daemons is more complicated (you first need to have them installed somewhere), but in general that could follow the same principle.

**Comment by WojciechowskiPiotr** [comment]
The `docker` library allows providing API version as `version` parameter so testing API should be quite easy. For changing the `docker` or `docker-py` we probably need a playbook to simplify a rotation via `pip`. Probably we can achieve a similar approach for rotating the docker version if we use it from a package but there might be system dependencies for the libraries.

**Comment by WojciechowskiPiotr** [comment]
Hello. I'm asking again for some feedback on my thought on the docker connection plugin I posted a week ago. I think that the plugin has serious limitations (like requirements of python installed inside of a container) and does not fit the way docker modules are working. At least some operations (copy to/from container) we may code as a module using the Docker API. @gundalow @felixfontein

**Comment by felixfontein** [comment]
I've spent some time looking at the `docker_image` module. Splitting it up into various modules is probably not so great, since that would require a lot of users to split their one `docker_image` task into several tasks. What do you think about the following changes to the interface instead?

- Add parameter `source` with options `build`, `load`, `pull` and `none`
  - Value will be autodetected (and show a warning that it will be required in Ansible 2.12)
  - Value `none` has no equivalent to current; will become default in Ansible 2.12; module will fail if image isn't present
  - Potential alternative to `none`: `local`. Any opinions?
- Rename options to make clear what they are used for:
  - `path` -> `build_path`
  - `rm` -> `build_rm`
  - `nocache` -> `build_nocache`
  - `http_timeout` -> `build_http_timeout`
  - `pull` -> `build_pull`
  - `dockerfile` -> `build_dockerfile`
  - `container_limits` -> `build_container_limits`
  - `buildargs` -> `build_args` (for consistency)
  - `cache_from` -> `build_cache_from`
  - `network_mode` -> `build_network_mode`
  - (Maybe also put these into a suboption? Like `build.xxx` instead of `build_xxx`?)
- Split up `force` parameter into different ones:
  - `force_source`: build/load/pull even though image is already present
  - `force_absent`: force image deletion
  - `force_tag`: force tagging
- Improve tag/push mechanism so that tagging actually checks if the destination image equals the image it should be (it currently doesn't check if the images are equal). Might need to be enabled by an option, like `tag_compare` (boolean, default `no`, if `yes` check whether image has changed)? This is more like a new feature, but I think a useful one, since most people don't really expect the tagging not being done just because the destination image already exists (but is totally unrelated to the current image).

**Comment by felixfontein** [comment]
FYI: the core and community freeze dates for Ansible 2.8 have been pushed back by a week (https://github.com/ansible/community/issues/346#issuecomment-473206145).

**Comment by felixfontein** [comment]
I have two `docker_image` PRs which need a review:
 - ansible/ansible#53911
 - ansible/ansible#53905

Can anyone help with them?

**Comment by felixfontein** [comment]
FYI: they have been moved again by a week, so Feature freeze is now April 4th.

**Comment by felixfontein** [comment]
FYI: it apparently has been moved again, so core freeze date is April 4th, and new feature freeze date is April 11th. There unfortunatley no official statement yet, so there's still some wiggle room, I'm pretty sure we at least have some extra days.

**Comment by felixfontein** [comment]
FYI: stable-2.8 has been branched. New features will have to wait for Ansible 2.9.

**Comment by felixfontein** [comment]
FYI: stable-2.8 has been branched. New features will have to wait for Ansible 2.9.

**Comment by felixfontein** [comment]
Congratulations to everyone involved in improving the docker modules and plugins for Ansible 2.8! I just went through the [changelog for Ansible 2.8 beta 1](https://github.com/ansible/ansible/blob/stable-2.8/changelogs/CHANGELOG-v2.8.rst), and there are *a lot* of docker-related changes! For the fun of it, I've extracted them all:

```.rst
Minor Changes
-------------

- The ``docker_image_facts`` module has been renamed to ``docker_image_info``.
- The ``docker_service`` module has been renamed to ``docker_compose``.
- The restart/idempotency behavior of docker_container can now be controlled with the new comparisons parameter.
- change default value for ``configs`` from ``[]`` to ``null`` and for ``update_order`` from ``stop-first`` to ``null``, matching docker API and allowing the module to interact with older docker daemons.
- docker_container - Add runtime option.
- docker_container - Add support for device I/O rate limit parameters. This includes ``device_read_bps``, ``device_write_bps``, ``device_read_iops`` and ``device_write_iops``
- docker_container - Added support for ``pids_limit`` parameter in docker_container.
- docker_container - Added support for healthcheck.
- docker_container - Allow to use image ID instead of image name.
- docker_container - ``stop_timeout`` is now also used to set the ``StopTimeout`` property of the docker container when creating the container.
- docker_container - a new option ``networks_cli_compatible`` with default value ``no`` has been added. The default value will change to ``yes`` in Ansible 2.12. Setting it to ``yes`` lets the module behave similar to ``docker create --network`` when at least one network is specified, i.e. the default network is not automatically attached to the container in this case.
- docker_container - improved ``diff`` mode to show output.
- docker_container - mount modes in ``volumes`` allow more values, similar to when using the ``docker`` executable.
- docker_container - published_ports now supports port ranges, IPv6 addresses, and no longer accepts hostnames, which were never used correctly anyway.
- docker_container, docker_network, docker_volume - return facts as regular variables ``container``, ``network`` respectively ``volume`` additionally to facts. This is now the preferred way to obtain results. The facts will be removed in Ansible 2.12.
- docker_image - Add ``build.cache_from`` option.
- docker_image - Allow to use image ID instead of image name for deleting images.
- docker_image - add option ``build.use_proxy_config`` to pass proxy config from the docker client configuration to the container while building.
- docker_image - all build-related options have been moved into a suboption ``build``. This affects the ``dockerfile``, ``http_timeout``, ``nocache``, ``path``, ``pull``, ``rm``, and ``buildargs`` options.
- docker_image - set ``changed`` to ``false`` when using ``force: yes`` to load or build an image that ends up being identical to one already present on the Docker host.
- docker_image - set ``changed`` to ``false`` when using ``force: yes`` to tag or push an image that ends up being identical to one already present on the Docker host or Docker registry.
- docker_image - the ``force`` option has been deprecated; more specific options ``force_source``, ``force_absent`` and ``force_tag`` have been added instead.
- docker_image - the ``source`` option has been added to clarify the action performed by the module.
- docker_image - the default for ``build.pull`` will change from ``yes`` to ``no`` in Ansible 2.12. Please update your playbooks/roles now.
- docker_image - the deprecated settings ``state: build`` and ``use_tls`` now display warnings when being used. They will be removed in Ansible 2.11.
- docker_image_facts - Allow to use image ID instead of image name.
- docker_network - Add support for IPv6 networks.
- docker_network - Minimum docker API version explcitly set to ``1.22``.
- docker_network - Minimum docker server version increased from ``1.9.0`` to ``1.10.0``.
- docker_network - Minimum docker-py version increased from ``1.8.0`` to ``1.10.0``.
- docker_network - ``attachable`` is now used to set the ``Attachable`` property of the docker network during creation.
- docker_network - ``internal`` is now used to set the ``Internal`` property of the docker network during creation.
- docker_network - ``scope`` is now used to set the ``Scope`` property of the docker network during creation.
- docker_network - add new option ``ipam_driver_options``.
- docker_network - added support for specifying labels
- docker_network - changed return value ``diff`` from ``list`` to ``dict``; the original list is contained in ``diff.differences``.
- docker_network - improved ``diff`` mode to show output.
- docker_secret - ``data`` can now accept Base64-encoded data via the new ``data_is_b64`` option. This allows to pass binary data or JSON data in unmodified form. (https://github.com/ansible/ansible/issues/35119)
- docker_service - return results as regular variable ``services``; this is a dictionary mapping service names to container dictionaries. The old ansible facts are still returned, but it is recommended to use ``register`` and ``services`` in the future. The facts will be removed in Ansible 2.12.
- docker_swarm - Added support for ``default_addr_pool`` and ``subnet_size``.
- docker_swarm - ``UnlockKey`` will now be returned when ``autolock_managers`` is ``true``.
- docker_swarm - module now supports ``--diff`` mode.
- docker_swarm_service - Add option ``limits`` as a grouper for resource limit options.
- docker_swarm_service - Add option ``logging`` as a grouper for logging options.
- docker_swarm_service - Add option ``placement`` as a grouper for placement options.
- docker_swarm_service - Add option ``reservations`` as a grouper for resource reservation options.
- docker_swarm_service - Add option ``restart_config`` as a grouper for restart options.
- docker_swarm_service - Add option ``update_config`` as a grouper for update options.
- docker_swarm_service - Added option ``resolve_image`` which enables resolving image digests from registry to detect and deploy changed images.
- docker_swarm_service - Added support for ``command`` parameter.
- docker_swarm_service - Added support for ``env_files`` parameter.
- docker_swarm_service - Added support for ``groups`` parameter.
- docker_swarm_service - Added support for ``healthcheck`` parameter.
- docker_swarm_service - Added support for ``hosts`` parameter.
- docker_swarm_service - Added support for ``rollback_config`` parameter.
- docker_swarm_service - Added support for ``stop_grace_period`` parameter.
- docker_swarm_service - Added support for ``stop_signal`` parameter.
- docker_swarm_service - Added support for ``working_dir`` parameter.
- docker_swarm_service - Added support for passing period as string to ``restart_policy_delay``.
- docker_swarm_service - Added support for passing period as string to ``restart_policy_window``.
- docker_swarm_service - Added support for passing period as string to ``update_delay``.
- docker_swarm_service - Added support for passing period as string to ``update_monitor``.
- docker_swarm_service - Extended ``mounts`` options. It now also accepts ``labels``, ``propagation``, ``no_copy``, ``driver_config``, ``tmpfs_size``, ``tmpfs_mode``.
- docker_swarm_service - ``env`` parameter now supports setting values as a dict.
- docker_swarm_service - added ``diff`` mode.
- docker_swarm_service: use docker defaults for the ``user`` parameter if it is set to ``null``
- docker_volume - changed return value ``diff`` from ``list`` to ``dict``; the original list is contained in ``diff.differences``.
- docker_volume - improved ``diff`` mode to show output.
- docker_volume - option minimal versions now checked. (https://github.com/ansible/ansible/issues/38833)
- docker_volume - reverted changed behavior of ``force``, which was released in Ansible 2.7.1 to 2.7.5, and Ansible 2.6.8 to 2.6.11. Volumes are now only recreated if the parameters changed **and** ``force`` is set to ``true`` (instead of or). This is the behavior which has been described in the documentation all the time.
- docker_volume - the ``force`` option has been deprecated, and a new option ``recreate`` has been added with default value ``never``. If you use ``force: yes`` in a playbook, change it to ``recreate: options-changed`` instead.
- inventory/docker - Group containers by docker-swarm "service" and "stack"

Deprecated Features
-------------------

- docker_network - Deprecate ``ipam_options`` in favour of ``ipam_config``.
- docker_swarm_service - Deprecate ``constraints`` in favour of ``placement``.
- docker_swarm_service - Deprecate ``limit_cpu`` and ``limit_memory`` in favour of ``limits``.
- docker_swarm_service - Deprecate ``log_driver`` and ``log_driver_options`` in favour of ``logging``.
- docker_swarm_service - Deprecate ``reserve_cpu`` and ``reserve_memory`` in favour of ``reservations``.
- docker_swarm_service - Deprecate ``restart_policy``, ``restart_policy_attempts``, ``restart_policy_delay`` and ``restart_policy_window`` in favour of ``restart_config``.
- docker_swarm_service - Deprecate ``update_delay``, ``update_parallelism``, ``update_failure_action``, ``update_monitor``, ``update_max_failure_ratio`` and ``update_order`` in favour of ``update_config``.

Bugfixes
--------

- docker connection - Support empty files with copying to target (https://github.com/ansible/ansible/issues/36725)
- docker_compose - fixed an issue where ``remove_orphans`` doesn't work reliably.
- docker_container - Fix idempotency problems with ``cap_drop`` and ``groups`` (when numeric group IDs were used).
- docker_container - Fix type conversion errors for ``log_options``.
- docker_container - Fixing various comparison/idempotency problems related to wrong comparisons. In particular, comparisons for ``command`` and ``entrypoint`` (both lists) no longer ignore missing elements during idempotency checks.
- docker_container - Makes ``blkio_weight``, ``cpuset_mems``, ``dns_opts`` and ``uts`` options actually work.
- docker_container - ``init`` and ``shm_size`` are now checked for idempotency.
- docker_container - ``publish_ports: all`` was not used correctly when checking idempotency.
- docker_container - do not fail when removing a container which has ``auto_remove: yes``.
- docker_container - fail if ``ipv4_address`` or ``ipv6_address`` is used with a too old docker-py version.
- docker_container - fail when non-string env values are found, avoiding YAML parsing issues. (https://github.com/ansible/ansible/issues/49802)
- docker_container - fix ``ipc_mode`` and ``pid_mode`` idempotency if the ``host:<container-name>`` form is used (as opposed to ``host:<container-id>``).
- docker_container - fix ``network_mode`` idempotency if the ``container:<container-name>`` form is used (as opposed to ``container:<container-id>``) (https://github.com/ansible/ansible/issues/49794)
- docker_container - fix ``paused`` option (which never worked).
- docker_container - fix behavior of ``detach: yes`` if ``auto_remove: yes`` is specified.
- docker_container - fix idempotency check for published_ports in some special cases.
- docker_container - fix idempotency problems with docker-py caused by previous ``init`` idempotency fix.
- docker_container - fix interplay of docker-py version check with argument_spec validation improvements.
- docker_container - fixing race condition when ``detach`` and ``auto_remove`` are both ``true``.
- docker_container - now returns warnings from docker daemon on container creation and updating.
- docker_container - refactored minimal docker-py/API version handling, and fixing such handling of some options.
- docker_container - the behavior is improved in case ``image`` is not specified, but needed for (re-)creating the container.
- docker_container, docker_image, docker_image_facts - also find local image when image name is prefixed with ``docker.io/library/`` or ``docker.io/``.
- docker_network - ``driver_options`` containing Python booleans would cause Docker to throw exceptions.
- docker_network - now returns warnings from docker daemon on network creation.
- docker_swarm - Fixed node_id parameter not working for node removal (https://github.com/ansible/ansible/issues/53501)
- docker_swarm - do not crash with older docker daemons (https://github.com/ansible/ansible/issues/51175).
- docker_swarm - fixes idempotency for the ``ca_force_rotate`` option.
- docker_swarm - improve Swarm detection.
- docker_swarm - improve idempotency checking; ``rotate_worker_token`` and ``rotate_manager_token`` are now also used when all other parameters have not changed.
- docker_swarm - now supports docker-py 1.10.0 and newer for most operations, instead only docker 2.6.0 and newer.
- docker_swarm - properly implement check mode (it did apply changes).
- docker_swarm - the ``force`` option was ignored when ``state: present``.
- docker_swarm_service - Added support for ``read_only`` parameter.
- docker_swarm_service - Document ``labels`` and ``container_labels`` with correct type.
- docker_swarm_service - Document ``limit_memory`` and ``reserve_memory`` correctly on how to specify sizes.
- docker_swarm_service - Document minimal API version for ``configs`` and ``secrets``.
- docker_swarm_service - Don't recreate service when ``networks`` parameter changes when running Docker API >= 1.29.
- docker_swarm_service - Don't set ``10`` as default for ``update_delay``.
- docker_swarm_service - Don't set ``1`` as default for ``update_parallelism``.
- docker_swarm_service - Don't set ``root`` as the default user.
- docker_swarm_service - Raise minimum required docker-py version for ``secrets`` to 2.4.0.
- docker_swarm_service - Raise minimum required docker-py version for module to 2.0.2.
- docker_swarm_service - Removed redundant defaults for ``uid``, ``gid``, and ``mode`` from ``configs`` and ``secrets``.
- docker_swarm_service - The ``publish``.``mode`` parameter was being ignored if docker-py version was < 3.0.0. Added a parameter validation test.
- docker_swarm_service - Validate choices for option ``mode``.
- docker_swarm_service - Validate minimum docker-py version of 2.4.0 for option ``constraints``.
- docker_swarm_service - When docker fails to update a container with an ``update out of sequence`` error, the module will retry to update up to two times, and only fail if all three attempts do not succeed.
- docker_swarm_service - fix use of Docker API so that services are not detected as present if there is an existing service whose name is a substring of the desired service
- docker_swarm_service - fixing falsely reporting ``publish`` as changed when ``publish.mode`` is not set.
- docker_swarm_service - fixing falsely reporting ``update_order`` as changed when option is not used.
- docker_swarm_service - fixing wrong option type for ``update_order`` which prevented using that option.
- docker_swarm_service - now returns warnings from docker daemon on service creation.
- docker_swarm_service - the return value was documented as ``ansible_swarm_service``, but the module actually returned ``ansible_docker_service``. Documentation and code have been updated so that the variable is now called ``swarm_service``. In Ansible 2.7.x, the old name ``ansible_docker_service`` can still be used to access the result.
- docker_swarm_service: fails because of default "user: root" (https://github.com/ansible/ansible/issues/49199)
- docker_volume - ``labels`` now work (and are a ``dict`` and no longer a ``list``).
- docker_volume - fix ``force`` and change detection logic. If not both evaluated to ``True``, the volume was not recreated.

New Plugins
-----------

Inventory
~~~~~~~~~

- docker_swarm - Ansible dynamic inventory plugin for Docker swarm nodes.

New Modules
-----------

Cloud
~~~~~

docker
^^^^^^

- docker_config - Manage docker configs.
- docker_container_info - Retrieves facts about docker container
- docker_host_info - Retrieves facts about docker host and lists of objects of the services.
- docker_network_info - Retrieves facts about docker network
- docker_node - Manage Docker Swarm node
- docker_node_info - Retrieves facts about docker swarm node from Swarm Manager
- docker_prune - Allows to prune various docker objects
- docker_stack - docker stack module
- docker_swarm_info - Retrieves facts about Docker Swarm cluster.
- docker_volume_info - Retrieve facts about Docker volumes
```

**Comment by WojciechowskiPiotr** [comment]
Top class work, everyone! :)

**Comment by blaisep** [comment]
Honestly, I feel lucky to be even remotely involved with this project!
Thanks.

**Comment by felixfontein** [comment]
If anyone's interested, we've started a discussion on extending image export/import functionality (by probably moving it into another module) in #54988.

**Comment by hannseman** [comment]
FYI Docker Python SDK 4.0.0 was just released. 
https://github.com/docker/docker-py/releases/tag/4.0.0

**Comment by felixfontein** [comment]
Ah, good news! Let's hope it doesn't break anything serious (like old code which assumes there's versions 1.x, 2.x and 3.x and nothing else, like it was until somewhen last year...) ;)

**Comment by felixfontein** [comment]
I've created #56615 to trigger all docker related tests. Let's see how that works out.

**Comment by hannseman** [comment]
@felixfontein nice! Looks like they passed just fine.

**Comment by felixfontein** [comment]
There are a couple of docker PRs which would benefit some reviews: #57913 #57914 (both general PRs, which improve error messages and error behavior), #58039 (docker_swarm_service bugfix), and #57911 (remove some sanity check ignores for docker code). Thanks! :)

**Comment by felixfontein** [comment]
Reminder: the Feature Freeze for Ansible 2.9 is coming up on [August 29th](https://docs.ansible.com/ansible/devel/roadmap/ROADMAP_2_9.html#expected), i.e. in less than two weeks! (And I guess/somewhat hope that this time it won't be delayed as much as last time.) So if you want to add something like a new feature, please hurry up :)

Small plug: ansible/ansible#60384 adds a mount collision check for docker_container and could use some reviews :)

**Comment by hannseman** [comment]
This PR could really use some thorough reviews:
https://github.com/ansible/ansible/pull/58961

**Comment by felixfontein** [comment]
There are two PRs which need reviews:
 - ansible/ansible#66599 (healthcheck disabling for docker_container and docker_swarm_service)
 - ansible/ansible#66382 (port ranges for docker_container revisited)

**Comment by felixfontein** [comment]
FYI, the devel branch is now frozen: https://groups.google.com/d/msg/ansible-devel/6UAfogEDtG8/wMTgxBGWCgAJ

**Comment by felixfontein** [comment]
We're currently discussing whether to move more stuff out of community.general into their own collection, and the docker modules and plugins are a prominent candidate: they could end up in a new collection community.docker. That would make development easier (people can actually have commit rights instead of having to work with the bot, you can subscribe to the repository instead of waiting for bot pings), would allow us to release when needed (like when we have bugfixes and/or new features) instead of waiting for the next community.general release, and to extend CI - like run the docker_swarm tests again! (This time it won't bother anyone else if they are unstable :) )

The discussion is taking place here: https://github.com/ansible-collections/overview/discussions/117#discussioncomment-106136

Feel free to voice your opinions/criticism/suggestions/... :)

**Comment by felixfontein** [comment]
If nobody complains, I'll start working on this probably on Wednesday.

**Comment by felixfontein** [comment]
(If you have any PRs that you would liked to get merged before that, please hurry up & ping me :) )

**Comment by WojciechowskiPiotr** [comment]
> (If you have any PRs that you would liked to get merged before that, please hurry up & ping me :) )

How much time does separation take? Will there be any period when we can't open PR? Or the approval process won't work correctly?

**Comment by felixfontein** [comment]
It should be done in at most 1-2 days. Then new PRs can be made in the new repository.

The modules will stay in community.general for some time (probably until the new collection is contained in an Ansible 2.10 release), if during that time someone creates a PR for one of the modules in community.general, I'll tell them to re-create it in the new repository (and help them if necessary).

**Comment by felixfontein** [comment]
I have a very first version of the repository active: https://github.com/ansible-collections/community.docker

Tests are mostly working: https://app.shippable.com/github/ansible-collections/community.docker/runs/8/summary/console The only thing not working yet is that I have to adjust all deprecations. I'll work on that this evening.

I plan to remove the community.general 2.0.0 removals right now, and change community.general 3.0.0 removals to community.docker 2.0.0.

I'll write here once I think the repository is in a good state.

**Comment by felixfontein** [comment]
Please check out ansible-collections/community.docker#1. That PR adjusts the deprecations and removes everything that was scheduled for community.general 2.0.0.

**Comment by felixfontein** [comment]
Please see ansible-collections/community.docker#4 for a release plan for community.docker.

**Comment by felixfontein** [comment]
@WojciechowskiPiotr I just saw that I never replied to your comments on the connection plugin(s) (now there are two :) ).

Python is only needed if you want to run modules (that include facts gathering) in the container. So if you use `gather_facts: false` and don't use modules, everything's fine. Unfortunately, even the `copy` module is not only an action plugin, but also a module, so effectively you can only use `raw` to execute commands. So yep, indeed, this isn't a really useful workaround if you don't happen to have Python inside the container. A separate module would be really helpful.

If anyone's interested in working on this, the `docker_api` connection plugin contains code for pushing files from and to the container using the API, these could be generalized and used by a new module.

**Comment by felixfontein** [comment]
If you have time, please take a look at the following PRs: ansible-collections/community.docker#87 ansible-collections/community.docker#89 ansible-collections/community.docker#90 ansible-collections/community.docker#93 Once these are merged, I plan to create the 1.3.0 release, which will contain a bugfix and a bunch of new features.

**Comment by WojciechowskiPiotr** [comment]
ansible-collections/community.docker#90 is still marked as WIP

**Comment by felixfontein** [comment]
It will stop being a WIP once 87 is merged, since it contains parts of 87.

**Comment by felixfontein** [comment]
(Doesn't mean it cannot be reviewed.)

**Comment by WojciechowskiPiotr** [comment]
I handled two, let others more familiar with `docker_image` look at the remaining

**Comment by felixfontein** [comment]
Would be great if someone can take a look at ansible-collections/community.docker#90 (docker_image_load) so we can release a new version by Monday (that will be included in Ansible 3.1.0, to be released on Tuesday).

**Comment by felixfontein** [comment]
I would be glad if someone could take a look at ansible-collections/community.docker#105, which adds a new module `docker_container_exec` which allows to run commands in docker containers.

**Comment by felixfontein** [comment]
I suggest we drop support for Ansible 2.9 and ansible-base 2.10 in the next major release. Both will be EOL in May 2022 anyway, and some versions of both cause problems (see https://github.com/ansible-collections/community.docker/pull/271). I think we should also have some policy when we drop support for ansible-core versions, and/or stop testing against them. I would suggest to strip EOL'ed ansible-core versions to minimal testing in CI (also so our CI matrix does not grow unbounded) and drop support for all EOL'ed ansible-core versions (and the ones which will EOL in the next 1-2 months) on every major release.

Does that sound reasonable? Any feedback, other ideas, ...?

(See also https://github.com/ansible-community/community-topics/issues/50 for a similar discussion for community.general and community.network.)

**Comment by felixfontein** [comment]
I'm wondering a bit whether Docker SDK for Python is finally dead, now that docker-compose switched from Python to Go. PRs and issues are being ignored, PR tests are not even triggered, there haven't been commits since mid-October 2021, ... There was a similar phase some years ago, which fortunately ended, but now with Docker SDK for Python no longer being used by Docker Inc. I'm somewhat worried.

**Comment by felixfontein** [comment]
Yes, that's the (official) Docker SDK for Python.

**Comment by felixfontein** [comment]
https://github.com/docker/docker-py/pull/2989 is a PR for an umaintained notice in Docker SDK for Python :disappointed:

**Comment by felixfontein** [comment]
I created an issue for discussing the future of the collection: https://github.com/ansible-collections/community.docker/issues/364

**Comment by felixfontein** [comment]
I'd like to deprecate Python 2.6 support: https://github.com/ansible-collections/community.docker/pull/388 Is anyone opposed to this?

**Comment by samccann** [comment]
@felixfontein since we will be archiving/freezing the ansible/community repo, can you move any further discussions like this to the forum and close out this issue?

**Comment by felixfontein** [comment]
Moving this to the forum doesn't make too much sense to me right now, since there's no tag / category for this (yet). I created a new topic in the community.crypto repository, https://github.com/ansible-collections/community.crypto/issues/24. Anyone interested in this topic, and general discussion about community.docker, should subscribe to that issue!

I'm closing this one now; there's no open discussion in here (in fact the last discussion has been some time ago). Thanks for everyone involved here who's no longer interested in following in the new topic, and I'm happy to see the ones that have continued interest in the collection in https://github.com/ansible-collections/community.docker/issues/701 or somewhere else instead :)

(For anyone who didn't knew yet: there's the new [Ansible forum](https://forum.ansible.com/), a discourse-based forum for the Ansible community. This can be used for support question, project discussion, information about Ansible related events, and anything else that's somehow related to Ansible. If you haven't already, please sign up! :) )
